File: lib/leetcode/121_best-time-to-buy-and-sell-stock.rb
# @param {Integer[]} prices
# @return {Integer}
def max_profit(stock_prices)
  return 0 if stock_prices.length < 2

  min_price = stock_prices[0]
  # first_profit = stock_prices[1] - stock_prices[0]
  max_profit = 0

  stock_prices[1..stock_prices.length].each do |current_price|
    potential_profit = current_price - min_price

    max_profit = [potential_profit, max_profit].max

    min_price = [current_price, min_price].min
  end

  max_profit
end

File: lib/leetcode/121_best-time-to-buy-and-sell-stock.js
/**
 * @param {number[]} prices
 * @return {number}
 */
const maxProfit = function(prices) {
  if (prices.length < 2) return 0;

  let minPrice = prices[0];
  let maxProfit = 0;

  for (let i = 1; i < prices.length; i++) {
    let currentProfit = prices[i] - minPrice;
    maxProfit = Math.max(currentProfit, maxProfit);
    minPrice = Math.min(minPrice, prices[i]);
  }

  return maxProfit;
};

let prices = [7, 1, 5, 3, 6, 4];
console.log(maxProfit(prices));
